js-api/src/chem.ts

```diff
@@ -114,73 +116,71 @@ export namespace chem {
     changedSub: Subscription | null = null;
     sketcher: SketcherBase | null = null;
     onChanged: Subject<any> = new Subject<any>();
-    sketcherCreated = new Subject<boolean>();
     sketcherFunctions: Func[] = [];
     selectedSketcher: Func | undefined = undefined;
-    extractorsCreated = new Subject<boolean>();
     sketcherDialogOpened = false;
 
     /** Whether the currently drawn molecule becomes the current object as you sketch it */
     syncCurrentObject: boolean = true;
 
     listeners: Function[] = [];
     _mode = SKETCHER_MODE.INPLACE;
-    _smiles = '';
-    _molfile = WHITE_MOLBLOCK;
-    _smarts = '';
-    unitsBeforeInit = '';
+    _smiles: string | null = null;
+    _molfile: string | null = null;
+    _smarts: string | null = null;
     molFileUnits = MOLV2000;
 
 
     extSketcherDiv = ui.div([], {style: {cursor: 'pointer'}});
     inplaceSketcherDiv: HTMLDivElement | null = null;
 
     getSmiles(): string {
-      return this.sketcher && this.sketcher._sketcher ? this.sketcher.smiles : this._smiles;
+      const smilesFromSmartsWarning = () => {
+        grok.shell.warning(`Smarts cannot be converted to smiles`);
+        return this._smarts!;
+      }
+      return this.sketcher && this.sketcher._sketcher ? this.sketcher.smiles : this._smiles === null ?
+        this._molfile !== null ? rdKitConvert(this._molfile, 'mol', 'smiles') :
+        this._smarts !== null ? smilesFromSmartsWarning() : '' : this._smiles;
     }
 
     setSmiles(x: string): void {
       this._smiles = x;
-      this.sketcher && this.sketcher._sketcher ? this.sketcher!.smiles = x : this.unitsBeforeInit = SMILES;
+      this._molfile = null;
+      this._smarts = null
+      if (this.sketcher && this.sketcher._sketcher) this.sketcher!.smiles = x;
       this.updateExtSketcherContent(this.extSketcherDiv);
     }
 
     getMolFile(): string {
       return this.sketcher && this.sketcher._sketcher ?
-        this.molFileUnits === MOLV2000 ? this.sketcher.molFile : this.sketcher.molV3000 : this._molfile;
+        this.molFileUnits === MOLV2000 ? this.sketcher.molFile : this.sketcher.molV3000 : this._molfile === null ?
```

Comment:
this.sketcher && this.sketcher._sketcher

Could be simplified as this.sketcher?._sketcher